### PR TITLE
Introduce size-based retention policy for soft-deletes

### PR DESCRIPTION
**Pull request by dnhatn**
Today soft-deleted documents retention policy is configured by an
absolute number of operations before the global checkpoint. Having
the reasonable number of operations for this setting is hard.

This commit replaces the operation-count policy with a size-based
policy. We translate the provided size to the number of operations
using the estimated average document size which is not perfect but
adequate over time.

Relates #34908

/cc @martijnvg and @jasontedor

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java [hunk 1]
```diff
@@ -102,9 +102,17 @@ synchronized long getMinRetainedSeqNo() {
             // then sends ops after the local checkpoint of that commit. This requires keeping all ops after localCheckpointOfSafeCommit;
             // - Changes APIs are driven the combination of the global checkpoint and retention ops. Here we prefer using the global
             // checkpoint instead of max_seqno because only operations up to the global checkpoint are exposed in the the changes APIs.
-            final long minSeqNoForQueryingChanges = globalCheckpointSupplier.getAsLong() - retentionOperations;
+            final long retentionOps;
+            if (retentionSizeInBytes > 0) {
+                // Use 1 byte per document if the average doc size is not available yet.
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java]: can you state when that is the case, when do we not have an avgDocSize? and in this case should we keep unlimited instead?

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 2]
```diff
@@ -246,10 +246,12 @@
...
      * Controls how many soft-deleted documents will be kept around before being merged away. Keeping more deleted
      * documents increases the chance of operation-based recoveries and allows querying a longer history of documents.
-     * If soft-deletes is enabled, an engine by default will retain all operations up to the global checkpoint.
+     * If soft-deletes is enabled, an engine by default will keep all soft-deleted documents in the most recent 4GB
+     * of Lucene index including live and soft-deleted documents. The extra required space depends on the deletion ratio.
      **/
-    public static final Setting<Long> INDEX_SOFT_DELETES_RETENTION_OPERATIONS_SETTING =
-        Setting.longSetting("index.soft_deletes.retention.operations", 0, 0, Property.IndexScope, Property.Dynamic);
+    public static final Setting<ByteSizeValue> INDEX_SOFT_DELETES_RETENTION_SIZE_SETTING =
+        Setting.byteSizeSetting("index.soft_deletes.retention.size", new ByteSizeValue(4, ByteSizeUnit.GB),
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java]: this looks like a bit too high for a default. What about 512MB or 1GB? @bleskes any gut feelings?

**Review by s1monw — COMMENTED** [review]
this LGTM I am a bit torn on the default and I think we need to add docs for this?

File: server/src/main/java/org/elasticsearch/common/lucene/Lucene.java [hunk 3]
```diff
@@ -159,6 +159,24 @@ public static int getNumDocs(SegmentInfos info) {
         return numDocs;
     }
 
+    /**
+     * Returns the average size in bytes of documents in the index referenced by a given {@link SegmentInfos}.
+     */
+    public static long getAverageDocumentSizeInBytes(SegmentInfos info) throws IOException {
+        long totalDocs = 0;
+        long totalBytes = 0;
+        for (SegmentCommitInfo si : info) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: is there an easy way to ignore tombstones and soft deletes here?

File: server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java [hunk 4]
```diff
@@ -95,16 +95,20 @@ private synchronized void releaseRetentionLock() {
...
         // Do not advance if the retention lock is held
-        if (retentionLockCount == 0) {
+        if (retentionLockCount == 0 && avgDocSizeInBytes > 0) {
             // This policy retains operations for two purposes: peer-recovery and querying changes history.
             // - Peer-recovery is driven by the local checkpoint of the safe commit. In peer-recovery, the primary transfers a safe commit,
             // then sends ops after the local checkpoint of that commit. This requires keeping all ops after localCheckpointOfSafeCommit;
             // - Changes APIs are driven the combination of the global checkpoint and retention ops. Here we prefer using the global
             // checkpoint instead of max_seqno because only operations up to the global checkpoint are exposed in the the changes APIs.
-            final long minSeqNoForQueryingChanges = globalCheckpointSupplier.getAsLong() - retentionOperations;
+            final long retentionOps = retentionSizeInBytes / avgDocSizeInBytes;
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java]: this gets weird if retentionSizeInBytes < 0 (specifically disabled - i.e., -1). Can we special case that?

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java (hunk 2 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java]: This is a good one. It's hard for me to come up with something good. Together with the time based policy we're evaluating it becomes easier (keep up to 24 hours but at most 5GB). Let's discuss this with the group?

File: server/src/main/java/org/elasticsearch/common/lucene/Lucene.java (hunk 3 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: why do you want to exclude soft-deletes?  I think it's ok to ignore the tombstones in this heuristic.

Comment (bleskes) [reply server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: I take your response to mean there is no easy way, which means I'm good with the heuristic as is. If there is an easy way to do it, I would consider it to protect against poisonous situations where a lot of index/delete patterns reduces doc size by 50%. Not a big deal, just a thought.

Comment (s1monw) [reply server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: no there is not except of iterating the entire index I guess. We can for instance only take segments into account with a low number of deleted documents to reduce the impact but I think we are ok here?

Comment (bleskes) [reply server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: I'm good thanks. Also, thinking more about this, one can argue that small tombstone docs actually should reduce the average size (as they are what we retain).

**Comment by bleskes** [comment]
@dnhatn if we move to this we need to make sure that the effective min required seq no is exposed in stats?

**Comment by dnhatn** [comment]
If we need a size based retention policy, we can revisit this PR later.

**Pull request closed without merging**